pandas/core/strings/accessor.py

```diff
@@ -1560,10 +1576,17 @@ def replace(
         if case is None:
             case = True
 
-        result = self._data.array._str_replace(
-            pat, repl, n=n, case=case, flags=flags, regex=regex
-        )
-        return self._wrap_result(result)
+        res_output = self._data
+        if not isinstance(pat, dict):
+            pat = {pat: repl}
+
+        for key, value in pat.items():
+            result = res_output.array._str_replace(
+                key, value, n=n, case=case, flags=flags, regex=regex
+            )
+            res_output = self._wrap_result(result)
```

Comment:
I believe you can call _wrap_result just once at the end, rather than inside the for loop.

Reply:
@rhshadrach wouldn't you need the for loop in the case that pat contained multiple key : value pairs of strings to be replaced?

Reply:
So after doing a bit of debugging it looks like we need to call _wrap_result after each iteration so we can save the output of our string replace and update `res_output`.

self._data is a Series and `_str_replace()` returns an NDArray. Since we can't update self._data.array, we need a container to save the output of our string replace, so we're converting it to a Series using `_wrap_result()` and then updating our container.

Reply:
I see - thanks for checking!

Reply:
rather than doing the loop here, is there any immediate advantage to passing the dict onto the arrays `_str_replace` method? like avoiding the `_wrap_result`

IIUC the accessors should only be validating the passed parameters, defining the "pandas string API", providing the documentation and wrapping the array result into a Series.

IMO the implementation should be at array level and then can be overridden if the array types can be optimized or use native methods.

For example, maybe using "._str_map" could be faster for object type and maybe `pyarrow.compute.replace_substring_regex` could be used for arrow backed strings?

The array level optimizations need not be in this PR though.

Reply:
> The array level optimizations need not be in this PR though.

I think this is a good idea, but agreed it need not be here (this is perf neutral compared to the status quo). If not tackled here, we can throw up an issue noting the performance improvement. @rmhowe425 - thoughts?

Reply:
Agreed! I think this idea is deserving of a separate issue. Happy to work that issue as well!